File: app/src/main/java/dev/radley/omgstarwars/text/CustomTypefaceSpan.java
package dev.radley.omgstarwars.text;

import android.graphics.Paint;
import android.graphics.Typeface;
import android.text.TextPaint;
import android.text.style.TypefaceSpan;

import org.jetbrains.annotations.NotNull;

/**
 * Enable multiple fonts in a textView
 *
 * Solution found on StackOverflow:
 * https://stackoverflow.com/questions/4819049/how-can-i-use-typefacespan-or-stylespan-with-a-custom-typeface
 *
 */
public class CustomTypefaceSpan extends TypefaceSpan {

    private final Typeface newType;

    /**
     *
     * @param family font family
     * @param type Typeface
     */
    public CustomTypefaceSpan(String family, Typeface type) {
        super(family);
        newType = type;
    }

    @Override
    public void updateDrawState(@NotNull TextPaint ds) {
        applyCustomTypeFace(ds, newType);
    }

    @Override
    public void updateMeasureState(@NotNull TextPaint paint) {
        applyCustomTypeFace(paint, newType);
    }

    private static void applyCustomTypeFace(Paint paint, Typeface tf) {
        int oldStyle;
        Typeface old = paint.getTypeface();
        if (old == null) {
            oldStyle = 0;
        } else {
            oldStyle = old.getStyle();
        }

        int fake = oldStyle & ~tf.getStyle();
        if ((fake & Typeface.BOLD) != 0) {
            paint.setFakeBoldText(true);
        }

        if ((fake & Typeface.ITALIC) != 0) {
            paint.setTextSkewX(-0.25f);
        }

        paint.setTypeface(tf);
    }
}

File: OMGStarWars/app/src/main/java/dev/radley/omgstarwars/text/CustomTypefaceSpan.kt
package dev.radley.omgstarwars.text

import android.graphics.Paint
import android.graphics.Typeface
import android.text.TextPaint
import android.text.style.TypefaceSpan

/**
 * Enable multiple fonts in a textView
 *
 * Solution found on StackOverflow:
 * https://stackoverflow.com/questions/4819049/how-can-i-use-typefacespan-or-stylespan-with-a-custom-typeface
 *
 */
class CustomTypefaceSpan(family: String, private val newType: Typeface?) : TypefaceSpan(family) {

    override fun updateDrawState(ds: TextPaint) {
        applyCustomTypeFace(ds)
    }

    override fun updateMeasureState(paint: TextPaint) {
        applyCustomTypeFace(paint)
    }

    private fun applyCustomTypeFace(paint: Paint) {

        newType?.let {
            val oldStyle: Int
            val old = paint.typeface

            oldStyle = old?.style ?: 0

            val fake = oldStyle and it.style.inv()
            if (fake and Typeface.BOLD != 0) {
                paint.isFakeBoldText = true
            }

            if (fake and Typeface.ITALIC != 0) {
                paint.textSkewX = -0.25f
            }

            paint.typeface = it
        }
    }
}

